instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/TripSessionsBillingTest.kt

```diff
@@ -0,0 +1,311 @@
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.setNavigationRoutesAndWaitForUpdate
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.startTripSessionAndWaitForActiveGuidanceState
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.startTripSessionAndWaitForFreeDriveState
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.stopTripSessionAndWaitForIdleState
+import com.mapbox.navigation.instrumentation_tests.utils.http.HttpServiceEvent
+import com.mapbox.navigation.instrumentation_tests.utils.http.HttpServiceEventsObserver
+import com.mapbox.navigation.instrumentation_tests.utils.parameters
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider.toNavigationRoutes
+import com.mapbox.navigation.testing.ui.BaseCoreNoCleanUpTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.testing.ui.utils.runOnMainSync
+import kotlinx.coroutines.delay
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.net.URI
+import java.util.concurrent.TimeUnit
+
+class TripSessionsBillingTest : BaseCoreNoCleanUpTest() {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+
+    private val route = RoutesProvider.dc_very_short(context).toNavigationRoutes()[0]
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        route.routeOptions.coordinatesList().first().run {
+            latitude = latitude()
+            longitude = longitude()
+        }
+        bearing = 190f
+    }
+
+    @Before
+    fun setup() {
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(context)
+                    .accessToken(getMapboxAccessTokenFromResources(context))
+                    .routingTilesOptions(
+                        RoutingTilesOptions.Builder()
+                            .tilesBaseUri(URI(mockWebServerRule.baseUrl))
+                            .build()
+                    )
+                    .build()
+            )
+        }
+    }
+
+    @Test
+    fun testBillingSessionSentForOneSecondLongActiveGuidanceTrip() = sdkTest {
+        val observer = HttpServiceEventsObserver.register()
+
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(listOf(route))
+        mapboxNavigation.startTripSession()
+
+        delay(TimeUnit.SECONDS.toMillis(SESSION_DURATION_SECONDS))
+
+        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()
+
+        delay(HTTP_EVENT_APPEARANCE_TIMEOUT_MILLIS)
```

Comment:
Are we playing with timeouts here? This could (and at some point will) make the tests flaky.
We could use other approaches here. Like waiting until the list has _n_ elements and fail by timeout. It would be safer. You can take a look at `NativeNavigatorCallbackOrderTest` and `waitUntilHasSize` method there (note that if you do this the list should be mutable, so what I said earlier about creating a copy in `HttpServiceUtils` would be irrelevant).

Reply:
Initially I written logic as you described: the test was waiting for n events, but this approach will return exactly n events even when there was more. And if we send more billing events than we should, it's a critical overbilling issue, I want to catch such cases. Unfortunately, `delay` won't totally guarantee that we catch such issues, but at least probability is higher. We agreed with @VysotskiVadim that this approach is better. @dzinad what do you think?

Reply:
During call, agreed with the team to change this behaviour: wait for exactly n event + add 1 second delay. This will allow us to avoid flakiness and still catch overbilling errors.